# Final Assignment for Week 3 of Coursera Capstone Project for Data Science

Since there are 3 parts to the final assignment, I will divide these notebook into 3 parts and complete it

In [1]:
import pandas as pd

## Part 1

In [2]:
import requests

In [3]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

This means there are no neighbourhoods in our table whose Borough is not assigned. Hooray!

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df.shape

(103, 3)

So there are 103 rows in our data set

## Part 2

In [14]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [15]:
import geocoder

import geocoder # import geocoder

#initialize your variable to None
 lat_lng_coords = None

postal_code = 'M5G'

#loop until you get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

The above was suggested in the assignment but because it was taking way too long to load, we will use the alternative method mentioned in the assignment

In [16]:
url = 'https://cocl.us/Geospatial_data'
data = pd.read_csv(url)
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [18]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [19]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [20]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [21]:
combined_data.shape

(103, 5)

That is the number of rows of the combined data table! We did it!

## Part 3

In [22]:
from geopy.geocoders import Nominatim

In [23]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


We will try to visualize the map of Toronto

In [28]:
!pip install folium

     |████████████████████████████████| 94 kB 5.5 MB/s  eta 0:00:01


In [29]:
import folium

In [35]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [31]:
CLIENT_ID = 'MRMDZXOFX3AUDPOADI5UZR3ZBMDQBL3OCRSNNSGEZIGVPMRV' 
CLIENT_SECRET = 'DW0WPEYU0XXTSDJ3NV3MNS4JM5YXQLJ2FNJQ14C4B3DMNQSW'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MRMDZXOFX3AUDPOADI5UZR3ZBMDQBL3OCRSNNSGEZIGVPMRV
CLIENT_SECRET:DW0WPEYU0XXTSDJ3NV3MNS4JM5YXQLJ2FNJQ14C4B3DMNQSW


In [33]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)


In [34]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [36]:
venues_in_toronto.shape

(1327, 5)

So we have got 1327 records. That's a lot for me. How about for you?

In [37]:
venues_in_toronto.head

<bound method NDFrame.head of                                           Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1322  Mimico NW, The Queensway West, South of Bloor,...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   
1325  Mimico NW, The Queensway West, South of Bloor,...   
1326  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.725882               -79.315572   
3                  43.725882     

In [38]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1314               43.628841               -79.520999              Wingporium   
1315               43.628841               -79.520999        South St. Burger   
1316               43.628841               -79.520999               Dollarama   
1317               43.628841               -79.520999          Healthy Planet   
1318               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1314            Wings Joint  
1315           Burger Joint  
1316         Discount Store  
1317        Supplement Shop  
1318      Convenience Store  

[417 rows x 5 columns]

Let's check for the maximum venue categories!

In [39]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[232 rows x 4 columns]

There are 232 different types of Venue Categories. Wow!

In [40]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1322                  0        0                   0             0   
1323                  0        0                   0             0   
1324                  0        0                   0             0   
1325                  0        0                   0             0   
1326                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1322               0                0                 0                    0   
1323               0                0                 0                    0   
1324               0                0                 0                    0   
1325               0                0                 0                    0   
1326               0                0                 0                    0   

      Antique Shop  Art Gallery  ...  Truck Stop  Turkish Restaurant  \
0                0            0  ...           0                   0   
1                0            0  ...           0                   0   
2                0            0  ...           0                   0   
3                0            0  ...           0                   0   
4                0            0  ...           0                   0   
...            ...          ...  ...         ...                 ...   
1322             0            0  ...           0                   0   
1323             0            0  ...           0                   0   
1324             0            0  ...           0                   0   
1325             0            0  ...           0                   0   
1326             0            0  ...           0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1322                              0                 0                      0   
1323                              0                 0                      0   
1324                              0                 0                      0   
1325                              0                 0                      0   
1326                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4  

Adding Neighbourhoods to the encoded dataframe

In [41]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  Turkish Restaurant  \
0                    0             0  ...           0                   0   
1                    0             0  ...           0                   0   
2                    0             0  ...           0                   0   
3                    0             0  ...           0                   0   
4                    0             0  ...           0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 233 columns]

Grouping the Neighbourhoods to calculate mean venue categories in each Neighbourhood

In [42]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0               0.0             0.000000           0.0  ...         0.0   
1               0.0             0.000000           0.0  ...         0.0   
2               0.0             0.000000           0.0  ...         0.0   
3               0.0             0.000000           0.0  ...         0.0   
4               0.0             0.041667           0.0  ...         0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 233 columns]

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [44]:
import numpy as np

Since there are too many neighbourhood categories, we will take the top 10 to cluster the neighbourhoods

In [45]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']


columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot        Clothing Store   
1                        Gym          Skating Rink              Pharmacy   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant                  Bank   Japanese Restaurant   
4         Italian Restaurant        Sandwich Place              Pharmacy   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0          Skating Rink                Deli / Bodega             Drugstore   
1                  Pool                          Pub        Sandwich Place   
2           Bridal Shop                Deli / Bodega        Sandwich Place   
3         Deli / Bodega  Eastern European Restaurant             Drugstore   
4             Juice Bar         Fast Food Restaurant      Sushi Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop               Dog Run    Distribution Center  
1           Coffee Shop   American Restaurant            Curling Ice  
2            Restaurant           Supermarket       Sushi Restaurant  
3            Donut Shop               Dog Run    Distribution Center  
4            Restaurant                   Pub           Liquor Store

Let's make the model to cluster the neighbourhoods

In [46]:
from sklearn.cluster import KMeans

In [48]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)


kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [49]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 2, 0, 1, 0, 1, 1, 1, 1, 1, 4, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1,
       1, 0, 1, 1, 1, 1, 0, 1], dtype=int32)

Adding the clustering Label column to the top 10 common venue categories

In [51]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Joining toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [52]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio            Curling Ice   
1           Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                  Park                  Café         Breakfast Spot   
3     Accessories Store           Event Space               Boutique   
4                 Diner      Sushi Restaurant            Yoga Studio   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0             Drugstore             Donut Shop               Dog Run   
1          Intersection                Dog Run            Donut Shop   
2                Bakery                Theater     French Restaurant   
3           Coffee Shop  Vietnamese Restaurant    Miscellaneous Shop   
4        Sandwich Place                   Park    Mexican Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0     Distribution Center        Discount Store                  Diner  
1     Distribution Center           Curling Ice                  Diner  
2   Performing Arts Venue        Chocolate Shop                    Pub  
3  Furniture / Home Store                 Diner           Dessert Shop  
4     Japanese Restaurant    Italian Restaurant    Fried Chicken Joint

Dropping all 'NaN' values to avoid data skew

In [53]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Let's plot the cluster on Map

In [54]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [55]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify our clusters

Cluster 1

In [56]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
52        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
66        North York             0.0                  Park   
83   Central Toronto             0.0            Playground   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio                  Curling Ice   
21         Women's Store          Dance Studio            Electronics Store   
35     Convenience Store          Intersection                  Yoga Studio   
52           Yoga Studio           Curling Ice  Eastern European Restaurant   
61           Swim School              Bus Line                  Yoga Studio   
66     Convenience Store           Yoga Studio                 Dance Studio   
83                  Park           Curling Ice  Eastern European Restaurant   
85                  Park          Intersection                  Curling Ice   
91            Playground                 Trail             Cuban Restaurant   

          5th Most Common Venue        6th Most Common Venue  \
0                     Drugstore                   Donut Shop   
21  Eastern European Restaurant                    Drugstore   
35                Deli / Bodega  Eastern European Restaurant   
52                    Drugstore                   Donut Shop   
61                Deli / Bodega  Eastern European Restaurant   
66  Eastern European Restaurant                    Drugstore   
83                    Drugstore                   Donut Shop   
85                    Drugstore                   Donut Shop   
91                    Drugstore                   Donut Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                Dog Run   Distribution Center        Discount Store   
21            Donut Shop               Dog Run   Distribution Center   
35             Drugstore            Donut Shop               Dog Run   
52               Dog Run   Distribution Center        Discount Store   
61             Drugstore            Donut Shop               Dog Run   
66            Donut Shop               Dog Run   Distribution Center   
83               Dog Run   Distribution Center        Discount Store   
85               Dog Run   Distribution Center        Discount Store   
91               Dog Run   Distribution Center        Discount Store   

   10th Most Common Venue  
0                   Diner  
21         Discount Store  
35    Distribution Center  
52                  Diner  
61    Distribution Center  
66         Discount Store  
83                  Diner  
85                  Diner  
91                  Diner

Cluster 2

In [57]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Pizza Place   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
7          North York             1.0                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto             1.0           Coffee Shop   
98          Etobicoke             1.0                  Park   
99   Downtown Toronto             1.0      Sushi Restaurant   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0                   Gym   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1             Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                    Park                  Café         Breakfast Spot   
3       Accessories Store           Event Space               Boutique   
4                   Diner      Sushi Restaurant            Yoga Studio   
7              Beer Store           Coffee Shop             Restaurant   
..                    ...                   ...                    ...   
97                   Café            Restaurant                  Hotel   
98                   Pool            Smoke Shop                  River   
99            Coffee Shop                   Pub               Beer Bar   
100         Auto Workshop           Pizza Place             Comic Shop   
102        Hardware Store                Bakery        Supplement Shop   

    5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
1            Intersection                Dog Run            Donut Shop   
2                  Bakery                Theater     French Restaurant   
3             Coffee Shop  Vietnamese Restaurant    Miscellaneous Shop   
4          Sandwich Place                   Park    Mexican Restaurant   
7    Caribbean Restaurant         Sandwich Place        Discount Store   
..                    ...                    ...                   ...   
97     Seafood Restaurant                    Gym              Tea Room   
98            Yoga Studio            Curling Ice            Donut Shop   
99                   Café      Indian Restaurant            Steakhouse   
100            Restaurant          Burrito Place            Skate Park   
102        Sandwich Place           Burger Joint     Convenience Store   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1       Distribution Center           Curling Ice                  Diner  
2     Performing Arts Venue        Chocolate Shop                    Pub  
3    Furniture / Home Store                 Diner           Dessert Shop  
4       Japanese Restaurant    Italian Restaurant    Fried Chicken Joint  
7                 Bike Shop   Sporting Goods Shop     Dim Sum Restaurant  
..                      ...                   ...                    ...  
97                Gastropub  Gym / Fitness Center            Pizza Place  
98                  Dog Run   Distribution Center         Discount Store  
99              Escape Room            Restaurant     Italian Restaurant  
100                 Brewery        Farmers Market   Fast Food Restaurant  
102              Kids Store  Fast Food Restaurant         Discount Store  

[87 rows x 12 columns]

Cluster 3

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
6          Dance Studio  Eastern European Restaurant             Drugstore   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6            Donut Shop               Dog Run   Distribution Center   

  9th Most Common Venue 10th Most Common Venue  
6        Discount Store                  Diner

Cluster 4

In [60]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
11   Distribution Center         Discount Store

Cluster 5

In [61]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
57   North York             4.0        Baseball Field   
101   Etobicoke             4.0        Breakfast Spot   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57                  Yoga Studio         Deli / Bodega     Electronics Store   
101  Construction & Landscaping        Baseball Field           Yoga Studio   

           5th Most Common Venue 6th Most Common Venue  \
57   Eastern European Restaurant             Drugstore   
101             Department Store     Electronics Store   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57                    Donut Shop               Dog Run   Distribution Center   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    10th Most Common Venue  
57          Discount Store  
101                Dog Run